lighthouse-core/audits/legacy-javascript.js

```diff
@@ -0,0 +1,360 @@
+   *
+   * @param {CodePatternMatcher} matcher
+   * @param {LH.GathererArtifacts['ScriptElements']} scripts
+   * @param {LH.Artifacts.NetworkRequest[]} networkRecords
+   * @return {Map<string, PatternMatchResult[]>}
+   */
+  static detectCodePatternsAcrossScripts(matcher, scripts, networkRecords) {
+    /** @type {Map<string, PatternMatchResult[]>} */
+    const urlToMatchResults = new Map();
+
+    for (const {requestId, content} of Object.values(scripts)) {
+      if (!content) continue;
+      const networkRecord = networkRecords.find(record => record.requestId === requestId);
+      if (!networkRecord) continue;
+      const matches = matcher.match(content);
+      if (!matches.length) continue;
+      urlToMatchResults.set(networkRecord.url, matches);
+    }
+
+    return urlToMatchResults;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const devtoolsLog = artifacts.devtoolsLogs[LegacyJavascript.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+    const mainResource = await MainResource.request({
+      URL: artifacts.URL,
+      devtoolsLog,
+    }, context);
+
+    /** @type {Array<{url: string, signals: string[], locations: LH.Audit.Details.SourceLocationValue[]}>} */
+    const tableRows = [];
+    let signalCount = 0;
+
+    // TODO(cjamcl): Use SourceMaps, and only pattern match if maps are not available.
+    const matcher = new CodePatternMatcher([
+      ...this.getPolyfillPatterns(),
+      ...this.getTransformPatterns(),
+    ]);
+
+    const urlToMatchResults =
+      this.detectCodePatternsAcrossScripts(matcher, artifacts.ScriptElements, networkRecords);
+    urlToMatchResults.forEach((matches, url) => {
+      /** @type {typeof tableRows[number]} */
+      const row = {url, signals: [], locations: []};
+      for (const match of matches) {
+        const {name, line, column} = match;
+        row.signals.push(name);
+        row.locations.push({
+          type: 'source-location',
+          url,
+          line,
+          column,
+          urlProvider: 'network',
+        });
+      }
+      tableRows.push(row);
+      signalCount += row.signals.length;
+    });
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headings = [
+      /* eslint-disable max-len */
+      {key: 'url', itemType: 'url', subRows: {key: 'locations', itemType: 'source-location'}, text: str_(i18n.UIStrings.columnURL)},
+      {key: '_', itemType: 'code', subRows: {key: 'signals'}, text: ''},
+      /* eslint-enable max-len */
+    ];
+    const details = Audit.makeTableDetails(headings, tableRows);
+
+    // Only fail if first party code has legacy code.
+    // TODO(cjamcl): Use third-party-web.
+    const foundSignalInFirstPartyCode = tableRows.some(row => {
+      return row.signals.length && URL.rootDomainsMatch(row.url, mainResource.url);
```

Comment:
I think this is a decent first pass for 3p filters 👍 

why would there be an entry in here that has no signals though?

also might be worth just filtering them out completely rather than just affecting the passing bit? if they're failing it's going to be annoying to see 90% unactionable things and just ignore the advice because they think they can't pass.

Reply:
Agreed - that's why I made the audit hide 3p by default in the report, for this audit. I don't much like the idea of _hiding_ insights - this audit should be able to tell you every bit of crufty JS on the page, for those inclined to know. For the rest (majority), hiding 3p by default prevents information overload.